### PR TITLE
hikey960: Fix race condition between hotplug and idles

### PR DESCRIPTION
**Pull request by Leo-Yan**
From the hotplug testing on Hikey960, in some case cores fail to become
online in the system.  When some cores are hotplugged off, if other
cores in the same cluster enter into CPU idle states at the meantime,
the cluster will be powered off.  This introduces the state machine
malfunction in the power controller, thus when hotplug on the core
afterwards, it fails to boot up the core because the power controller
thinks the cluster is powered on.

This patch is to avoid race condition between hotplug and idles by
preventing cluster power off when some of cores in the cluster are
hotplugged off, if all cores in the same cluster are hotplugged off,
the cluster can be powered off.

Signed-off-by: Wei Yu <yuwei3@hisilicon.com>
Signed-off-by: Leo Yan <leo.yan@linaro.org>

### CODE REVIEW

**Comment by Leo-Yan** [comment]
@hzhuang1 , could you give a review for this patch, thanks!

**Comment by hzhuang1** [comment]
acked

File: plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c [hunk 1]
```diff
@@ -268,7 +286,8 @@ int hisi_test_pwrdn_allcores(unsigned int cluster, unsigned int core)
 	mask = (PDC_COREPWRSTAT_MASK & (~mask));
 	pdc_stat &= mask;
 
-	if ((boot_flag ^ cpuidle_flag) || pdc_stat)
+	if ((boot_flag ^ cpuidle_flag) || pdc_stat ||
+	    heck_hotplug(cluster, boot_flag))
```
Comment (JoelHutton) [inline plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c]: should this be 'check_hotplug'? does this build?

**Comment by Leo-Yan** [comment]
Hi @JoelHutton, thanks for reviewing.  It's my bad when formatting the patch and now have uploaded the new patch.

**Comment by dp-arm** [comment]
IP Review +1

File: plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c [hunk 2]
```diff
@@ -149,11 +149,17 @@ void hisi_clear_cpuidle_flag(unsigned int cluster, unsigned int core)
...
-	val &= AP_SUSPEND_FLAG;
-	return !!val;
+	val1 = mmio_read_32(CPUIDLE_FLAG_REG(0));
+	val1 &= AP_SUSPEND_FLAG;
+
+	val2 = mmio_read_32(CPUIDLE_FLAG_REG(1));
+	val2 &= AP_SUSPEND_FLAG;
+
+	val1 |= val2;
+	return !!val1;
```
Comment (JoelHutton) [inline plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c]: I personally find the double negation less readable.
perhaps:
return (val1 == 1);

Comment (JoelHutton) [reply plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c]: sorry, I meant:
return (val1 != 0);

**Comment by valschneider** [comment]
Hi,

I gave this a spin on my H960, but the issue still seems to be here (although slightly harder to trigger).

I'm mostly clueless about ATF but it does seem like I got the right firmware in there (I followed the instructions at https://github.com/96boards-hikey/tools-images-hikey960/blob/master/build-from-source/README-ATF-UEFI-build-from-source.md):
```
NOTICE:  BL2: v2.0(release):v2.0-1052-gd8c6c336
NOTICE:  BL2: Built : 15:58:53, Mar 20 2019
NOTICE:  BL2: Booting BL31
NOTICE:  BL31: v2.0(release):v2.0-1052-gd8c6c336
NOTICE:  BL31: Built : 15:58:53, Mar 20 2019
UEFI firmware (version Alpha built at 15:29:23 on Mar 20 2019)
```
This matches d8c6c3366fa8 ("hikey960: Fix race condition between hotplug and idles")



As for the dmesg errors, this is what I get:
```
[   40.967321] CPU1: shutdown
[   40.970064] psci: CPU1 killed.
[   42.079800] CPU1: failed to come online
[   42.083711] CPU1: failed in unknown state : 0x0
sh: write error: Input/output error

[  148.464746] psci: failed to boot CPU1 (-22)
[  148.469059] CPU1: failed to boot: -22
-bash: echo: write error: Invalid argument
```

The test is 10 iterations of LISA's hotplug tests with:
```
exekall run lisa/tests/ --conf /data/work/platforms/hikey960/debian/target_conf.yml -n 10 -s HotplugTorture* --share '*Target'
```

This runs 10 iterations of that script:
(the sleep durations are random, but this particular script seems to have the right timing)
```bash
#!/root/devlib-target/bin/busybox sh
set -e
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.085
echo 0 > /sys/devices/system/cpu/cpu5/online
sleep 0.08
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.018
echo 0 > /sys/devices/system/cpu/cpu3/online
sleep 0.05
echo 0 > /sys/devices/system/cpu/cpu0/online
sleep 0.049
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.064
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.037
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.033
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.025
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.018
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.069
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.036
echo 0 > /sys/devices/system/cpu/cpu4/online
sleep 0.06
echo 1 > /sys/devices/system/cpu/cpu3/online
sleep 0.067
echo 1 > /sys/devices/system/cpu/cpu4/online
sleep 0.055
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.052
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.035
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.053
echo 0 > /sys/devices/system/cpu/cpu4/online
sleep 0.081
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.092
echo 1 > /sys/devices/system/cpu/cpu0/online
sleep 0.053
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.041
echo 1 > /sys/devices/system/cpu/cpu4/online
sleep 0.043
echo 1 > /sys/devices/system/cpu/cpu5/online
sleep 0.012
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.077
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.038
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.092
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.07
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.05
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.092
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.077
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.061
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.075
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.072
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.097
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.059
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.019
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.09
echo 0 > /sys/devices/system/cpu/cpu5/online
sleep 0.033
echo 1 > /sys/devices/system/cpu/cpu5/online
sleep 0.018
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.014
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.013
echo 0 > /sys/devices/system/cpu/cpu5/online
sleep 0.024
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.018
echo 1 > /sys/devices/system/cpu/cpu5/online
sleep 0.087
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.091
echo 0 > /sys/devices/system/cpu/cpu4/online
sleep 0.051
echo 0 > /sys/devices/system/cpu/cpu0/online
sleep 0.035
echo 0 > /sys/devices/system/cpu/cpu3/online
sleep 0.051
echo 0 > /sys/devices/system/cpu/cpu5/online
sleep 0.062
echo 1 > /sys/devices/system/cpu/cpu5/online
sleep 0.057
echo 1 > /sys/devices/system/cpu/cpu0/online
sleep 0.013
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.034
echo 0 > /sys/devices/system/cpu/cpu0/online
sleep 0.03
echo 0 > /sys/devices/system/cpu/cpu5/online
sleep 0.092
echo 1 > /sys/devices/system/cpu/cpu0/online
sleep 0.071
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.057
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.097
echo 0 > /sys/devices/system/cpu/cpu0/online
sleep 0.05
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.073
echo 1 > /sys/devices/system/cpu/cpu5/online
sleep 0.095
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.099
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.014
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.034
echo 1 > /sys/devices/system/cpu/cpu0/online
sleep 0.054
echo 1 > /sys/devices/system/cpu/cpu3/online
sleep 0.063
echo 1 > /sys/devices/system/cpu/cpu4/online
sleep 0.02
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.045
echo 0 > /sys/devices/system/cpu/cpu4/online
sleep 0.1
echo 0 > /sys/devices/system/cpu/cpu5/online
sleep 0.033
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.028
echo 0 > /sys/devices/system/cpu/cpu0/online
sleep 0.02
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.057
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.028
echo 0 > /sys/devices/system/cpu/cpu3/online
sleep 0.078
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.095
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.038
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.06
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.098
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.037
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.038
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.013
echo 1 > /sys/devices/system/cpu/cpu3/online
sleep 0.082
echo 0 > /sys/devices/system/cpu/cpu3/online
sleep 0.012
echo 1 > /sys/devices/system/cpu/cpu3/online
sleep 0.074
echo 0 > /sys/devices/system/cpu/cpu6/online
sleep 0.074
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.038
echo 1 > /sys/devices/system/cpu/cpu6/online
sleep 0.034
echo 1 > /sys/devices/system/cpu/cpu4/online
sleep 0.094
echo 0 > /sys/devices/system/cpu/cpu1/online
sleep 0.081
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.05
echo 0 > /sys/devices/system/cpu/cpu2/online
sleep 0.08
echo 0 > /sys/devices/system/cpu/cpu4/online
sleep 0.099
echo 1 > /sys/devices/system/cpu/cpu5/online
sleep 0.083
echo 1 > /sys/devices/system/cpu/cpu4/online
sleep 0.08
echo 0 > /sys/devices/system/cpu/cpu7/online
sleep 0.022
echo 1 > /sys/devices/system/cpu/cpu0/online
sleep 0.033
echo 1 > /sys/devices/system/cpu/cpu1/online
sleep 0.051
echo 1 > /sys/devices/system/cpu/cpu2/online
sleep 0.065
echo 1 > /sys/devices/system/cpu/cpu7/online
sleep 0.018
set +e
```

**Comment by Leo-Yan** [comment]
Hi @valschneider,

Thanks a lot for testing! 

At my side, I used my testing script for running 40+ minutes and cannot reproduce the hotplug stuck issue;  and I also tried your shared script for the testing for running 20+ iterations and don't see the hotplug failure.

My testing script:
~~~
while true; do
	echo 0 > /sys/devices/system/cpu/cpu4/online
	sleep 1
	echo 0 > /sys/devices/system/cpu/cpu5/online
	sleep 1
	echo 0 > /sys/devices/system/cpu/cpu6/online
	sleep 1
	echo 0 > /sys/devices/system/cpu/cpu7/online
	sleep 1

	echo 1 > /sys/devices/system/cpu/cpu4/online
	sleep 1
	echo 1 > /sys/devices/system/cpu/cpu5/online
	sleep 1
	echo 1 > /sys/devices/system/cpu/cpu6/online
	sleep 1
	echo 1 > /sys/devices/system/cpu/cpu7/online
	sleep 1
done
~~~

Could you test my shared fip.bin: http://people.linaro.org/~leo.yan/hikey960/fip.bin?  Since fip.bin contains UEFI and MCU firmware, I want to confirm if my built binary can work at your side or not?  I built it with latest code bases for all repositories.

If you still see the failure, I will sync with you offline to check kernel version (I am using mainline kernel + Debian) and LISA testing scripts.  Very appreciate your help.

File: plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c (hunk 2 above)
Comment (Leo-Yan) [reply plat/hisilicon/hikey960/drivers/pwrc/hisi_pwrc.c]: Yes, will change to this way (and upload the new patch after settle down the issue reported by Valentine).

**Comment by antonio-nino-diaz-arm** [comment]
I feel that if this improves stability we could merge it as long as the commit message is updated to reflect the fact that this is not a complete fix.

**Comment by Leo-Yan** [comment]
Hi @antonio-nino-diaz-arm, just head up that Wei has worked out a new patch and we are testing on this.  So if you think it's not urgent, I'd like to suggest to hold on to see what's the testing result with the new patch.

**Comment by antonio-nino-diaz-arm** [comment]
Ok, we can wait then.

**Comment by antonio-nino-diaz-arm** [comment]
As part of the migration to trustedfirmware.org https://github.com/ARM-software/tf-issues/issues/681 we have switched the main development repository to Gerrit (https://review.trustedfirmware.org/). It is unfortunate that the switch has happened while your PR was being reviewed, but we don't intend to merge more patches through GitHub to avoid confusion.

Could you please follow the instructions in https://github.com/ARM-software/arm-trusted-firmware/blob/master/contributing.rst to continue with the review? We're sorry for the inconvenience.

**Comment by sandrine-bailleux-arm** [comment]
In addition to the [contribution guidelines](https://github.com/ARM-software/arm-trusted-firmware/blob/master/contributing.rst), the following [Getting starting guide](https://developer.trustedfirmware.org/w/tf_a/gerrit-getting-started/) might help in this process.

**Comment by Leo-Yan** [comment]
Hi @antonio-nino-diaz-arm , @sandrine-bailleux-arm, thanks for the info for the new process for using gerrit for reviewing.  I have created account and upload the new patch on the gerrit:
https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/819

Hi @hzhuang1, @valschneider, since you have not created account on gerrit so I cannot add your into reviewer list.  You could review it with the link.  I did the testing with Valentin's script with new patch, but unfortunately I still can see CPU0 cannot hotplug back, but big CPUs hotplug on/off issue have been resolved. If you think it's okay, I think we can merge the patch firstly and I will check with Wei for CPU0 hotplug on/off failure issue.  Please let me know what's you think about this.  Thanks!

**Comment by hzhuang1** [comment]
I register it now. But I'm curious on the variable of "cluster". Does it mean that these two clusters bind closely?

**Comment by Leo-Yan** [comment]
> I register it now. But I'm curious on the variable of "cluster". Does it mean that these two clusters bind closely?

I don't think these two clusters are bound together for power management, but from the point of view of registers design, the MCU uses the same register (or same word in the SRAM) to serve the requests for two clusters.

On the other words, for the hardware design, the clock/power/voltage are separated for two clusters; but the MCU doesn't use very distinguished register layout to present this.

**Comment by antonio-nino-diaz-arm** [comment]
We'll continue the review here: https://review.trustedfirmware.org/#/c/TF-A/trusted-firmware-a/+/819/

**Pull request closed without merging**